src/main/java/org/opentripplanner/routing/graphfinder/PlaceFinderTraverseVisitor.java

```diff
@@ -168,16 +175,31 @@ private boolean stopHasPatternsWithMode(RegularStop stop, Set<TransitMode> modes
   }
 
   private void handleStop(RegularStop stop, double distance) {
-    if (filterByStops != null && !filterByStops.contains(stop.getId())) {
+    if (
+      (!filterByStops.isEmpty() && !filterByStops.contains(stop.getId())) ||
+      seenStops.contains(stop.getStationOrStopId()) ||
+      !(filterByModes.isEmpty() || stopHasPatternsWithMode(stop, filterByModes))
+    ) {
```

Comment:
This is a bit hard to read. I think there are multiple ways to solve this issue. Either add some comment, split the check into multiple ifs and/or hide the checks behind some functions with clear names.

Reply:
This should be addressed now - extracted two functions and added a comment